### PR TITLE
Remove over meta programming in AR::Relation

### PR DESCRIPTION
**Pull request by bogdan**
### Summary

Introduced low level methods #set and #get for setting query attributes:

  relation.set(:where, {id: 1})
  relation.get(:includes)

Used those internally when working with relation's attributes
at the abstract level

### CODE REVIEW

**Comment by rafaelfranca** [comment]
There are some style violations in the PR. The implementation is good to me, could you fix the violations?

**Comment by matthewd** [comment]
I don't really like the names `get` and `set`, but I'm struggling to offer better suggestions. We should at least consider prepending an underscore -- relation is a namespace we share with user methods, so short + generic + undifferentiated seems conflict-prone.

**Comment by bogdan** [comment]
@matthewd from the other hand I sometimes would want to use `get` and `set` in the rails app instead of not so convenient and strange accessor (especially `get`).  So I want to treat them as official `Relation` API.

**Comment by rafaelfranca** [comment]
Well, it will be private API so it can't be official `Relation`.

**Comment by rafaelfranca** [comment]
`get_property`/`set_property` maybe?

**Comment by bogdan** [comment]
Ok, I'll gather all variants for you:

* `_get`
* `__get__`
* `get`
* `get_property`
* `[]=`
* `get_value`
* `relation_value=`

Choose wisely.

**Comment by bogdan** [comment]
@rafaelfranca changed to `set/get_value` as they seem the most relevant to current terminology of "values" inside relation.

File: activerecord/lib/active_record/relation/query_methods.rb [hunk 1]
```diff
@@ -56,49 +56,25 @@ def not(opts, *rest)
...
+      method_name = case name
+                    when *Relation::MULTI_VALUE_METHODS then "#{name}_values"
+                    when *Relation::SINGLE_VALUE_METHODS then "#{name}_value"
+                    when *Relation::CLAUSE_METHODS then "#{name}_clause"
+      end
       class_eval <<-CODE, __FILE__, __LINE__ + 1
-        def #{name}_value                    # def readonly_value
-          @values[:#{name}]                  #   @values[:readonly]
+        def #{method_name}                   # def includes_values
+          get_value(#{name.inspect})               #   get_value(:includes)
```
Comment (kamipo) [inline activerecord/lib/active_record/relation/query_methods.rb]: Broken an alignment of comments.

File: activerecord/lib/active_record/relation/query_methods.rb [hunk 2]
```diff
@@ -56,49 +56,25 @@ def not(opts, *rest)
...
-      CODE
-    end
 
-    Relation::SINGLE_VALUE_METHODS.each do |name|
-      class_eval <<-CODE, __FILE__, __LINE__ + 1
-        def #{name}_value=(value)            # def readonly_value=(value)
-          assert_mutability!                 #   assert_mutability!
-          @values[:#{name}] = value          #   @values[:readonly] = value
+        def #{method_name}=(value)           # def includes_values=(value)
+          set_value(#{name.inspect}, value)        #   set_value(:includes, value)
```
Comment (kamipo) [inline activerecord/lib/active_record/relation/query_methods.rb]: ditto.

File: activerecord/lib/active_record/relation/query_methods.rb [hunk 3]
```diff
@@ -950,6 +924,17 @@ def arel # :nodoc:
       @arel ||= build_arel
     end
 
+    # Returns a relation value with a given name
+    def get_value(name) # :nodoc:
+      @values[name] || default_value_for(name)
+    end
+
+    # Sets the relation value with the given name
+    def set_value(name, value) #:nodoc:
```
Comment (kamipo) [inline activerecord/lib/active_record/relation/query_methods.rb]: I am a little curious about `#:nodoc:`. `get_value` is `# :nodoc:` (including a space).

**Comment by kamipo** [comment]
another variants: `read_value`/`write_value`.

File: activerecord/lib/active_record/relation/query_methods.rb [hunk 4]
```diff
@@ -418,7 +389,10 @@ def unscope!(*args) # :nodoc:
       args.each do |scope|
         case scope
         when Symbol
-          symbol_unscoping(scope)
+          if !VALID_UNSCOPING_VALUES.include?(scope)
+            raise ArgumentError, "Called unscope() with invalid unscoping argument ':#{scope}'. Valid arguments are :#{VALID_UNSCOPING_VALUES.to_a.join(", :")}."
+          end
+          set_value(scope, nil)
```
Comment (kamipo) [inline activerecord/lib/active_record/relation/query_methods.rb]: Is setting `nil` safe for `MULTI_VALUE_METHODS`?

Comment (bogdan) [inline activerecord/lib/active_record/relation/query_methods.rb]: It results by overwriting by default on line 929

Comment (kamipo) [inline activerecord/lib/active_record/relation/query_methods.rb]: I see. Thanks for your explanation!

**Comment by bogdan** [comment]
@rafaelfranca @matthewd ready to merge?

**Pull request merged**